### ISSUE TITLE
[SEC] Eth2.0 Req/Resp - input validation for beaconBlocksByRoot request handler

### ISSUE THREAD

**Issue by tintinweb** [status-im/nimbus-eth2#1537]
## Description
This is basically an addition to https://github.com/status-im/nim-beacon-chain/issues/1366 addressing input validation issues in the `beaconBlocksByRoot` rpc request handler.

For example, the request handler is not validating that the request to provide blocks for roots complies with the specification.

According to the [specification](https://github.com/ethereum/eth2.0-specs/blob/2a41c516d302b08784dea5f61c95dd905829020f/specs/phase0/p2p-interface.md#BeaconBlocksByRoot)

> No more than MAX_REQUEST_BLOCKS may be requested at a time.

However, the request handler does not verify the number of blocks requested and may be tricked into providing more than `MAX_REQUEST_BLOCKS` blocks. Furthermore, the method does not fail if a peer is sending a request for an empty list of roots.

## Exploit Scenario

* also see https://github.com/status-im/nim-beacon-chain/issues/1366 
* spec violation
* depending on how resource intensive the lookup and response are this may be used for DoS attacks (client asking for way more blocks than acceptable by spec or client)
* requests for zero blocks do not fail or penalize the peer. depending on how the scoring system is set up, valid requests may even gain a positive score. In such a case, sending a lot of empty requests may be used by a peer to push its own score (note: peer scoring is not yet implemented for the rpc responding side: https://github.com/status-im/nim-beacon-chain/issues/1365).

## Mitigation Recommendation
* strictly validate rpc requests according to be within bound of the spec
* disallow and penalize requests containing zero or greater than the specifications `MAX_REQUEST_BLOCKS` block requests

## References
https://github.com/status-im/nim-beacon-chain/blob/3433c77c35f3025729ac9204ab6b91ea136dce9e/beacon_chain/sync_protocol.nim.generated.nim#L220-L240

**Comment by zah** [comment]
This is a bit tricky to spot, but there is input validation here provided by the `BlockRootsList` type which is used as a parameter:

```nim
type
  BlockRootsList* = List[Eth2Digest, Limit MAX_REQUEST_BLOCKS]
```

This is a SSZ List type with a limit enforced during deserialization here:

https://github.com/status-im/nim-beacon-chain/blob/master/beacon_chain/ssz/bytes_reader.nim#L18-L21
https://github.com/status-im/nim-beacon-chain/blob/master/beacon_chain/ssz/bytes_reader.nim#L160

In other words, when the incoming message violates the limit, the attempt to deserialize the parameter will result in an exception that is handled here:

https://github.com/status-im/nim-beacon-chain/blob/master/beacon_chain/eth2_network.nim#L638-L639

So, we will respond with `InvalidRequest` which is in accordance to the spec.

**Comment by zah** [comment]
The suggestion to penalized empty requests will be considered.

**Comment by tintinweb** [comment]
I definitely missed that, thanks @zah 🙌 

Do you want me to re-write the issue or should we leave it as is for future reference? just in case someone is looking for the same thing?

**Comment by zah** [comment]
It's up to you. My preference would be to leave you to what you do best - finding these issues :)

**Comment by arnetheduck** [comment]
Sounds like the explanation part is something we should put in a code comment

**Comment by tintinweb** [comment]
invalid -> closed

- inline comment added 🙌 
- allows empty request (will only result in a log message; cannot be used to improve peers score as scoring is not adjusted for rpc requests)

**Comment by zah** [comment]
There have been some additional developments here which I have forgotten to link properly to the issue:

In cca25e843b97c405096c4e255596aced457f48d6, I've added a new exception called `InvalidInputsError` which is now raised properly on empty requests:

https://github.com/status-im/nimbus-eth2/blob/cca25e843b97c405096c4e255596aced457f48d6/beacon_chain/sync_protocol.nim#L137-L171

https://github.com/status-im/nimbus-eth2/blob/cca25e843b97c405096c4e255596aced457f48d6/beacon_chain/sync_protocol.nim#L173-L181

It results in the connection being terminated with the `InvalidRequest` status code.

**Comment by tintinweb** [comment]
ah okay, these changes are in the unmerged misc-fixes branch. 👍
